### PR TITLE
Do not assume that the packet will not be an ACK in NlSocketHandle::recv

### LINKED ISSUE #134: nl80211 does not always sends data.
Some nl80211 commands does only return a Ack, and no data messages. This breaks the current assumptions of NlSocketHandle.

### PR DESCRIPTION
**Pull request by jbaublitz**
Closes #134

### CODE REVIEW

File: src/socket.rs [hunk 1]
```diff
@@ -562,30 +562,39 @@ impl NlSocketHandle {
...
-            {
-                return Err(NlError::NoAck);
+            Err(NlError::from(e))
+        } else if let NlPayload::Ack(_) = packet.nl_payload {
+            if self.needs_ack {
+                self.needs_ack = false;
+                Ok(None)
+            } else {
+                Err(NlError::new(
+                    "Socket did not expect an ACK but one was received",
```
Comment (Kloenk) [inline src/socket.rs:572]: Shouldn't this be something like NlError::NoAckWanted

**Comment by Kloenk** [comment]
This breaks `socket.resolve_genl_family("nl80211")?`. It creates "netlink error: Socket did not expect an ACK but one was received"

**Comment by Kloenk** [comment]
Just setting `self.needs_ack = true;` after [this line](https://github.com/jbaublitz/neli/blob/issue-134/src/socket.rs#L371) should probably fix it

File: src/socket.rs (hunk 1 above)
Comment (jbaublitz) [reply src/socket.rs:572]: It should be, but that would be a breaking change for anyone matching on the enum. I'll put up an issue for that in 0.6.0.

**Comment by jbaublitz** [comment]
> Just setting `self.needs_ack = true;` after [this line](https://github.com/jbaublitz/neli/blob/issue-134/src/socket.rs#L371) should probably fix it

I'm going to dig into this further because that should already be set in the implementation of `NlSocketHandle::send()` so obviously something is broken.

**Comment by Kloenk** [comment]
I thing it's only in send_nl and not in send. I think send is a more low level api

**Comment by jbaublitz** [comment]
`send_nl` has been renamed to `send` in `NlSocketHandler` so it should be set.

**Comment by Kloenk** [comment]
Maybe this is there, because it does not use recv, but iter? Not sure, not used the iter function yet, just saw that the resolve family function uses it.

**Comment by jbaublitz** [comment]
@Kloenk I'm sorry for the delay. I've primarily been working on #143. I'm shifting focus back to this and I wanted to propose something.

I'm counting the current implementation as a bug and so I'm willing to introduce a breaking change in a patch version if the current solution does not work.

What I'm currently considering is:
* `NlSocketHandle::recv` should only return `None` if the socket is non-blocking
* `NlMessageIter` should handle the calculation of `None` for the user
* `NlSockHandle::recv` should return _all_ packets to the user without filtering out ACKS (I will even eventually need this for #104)

How does that sound to you?

**Comment by Kloenk** [comment]
I never worked with `NlMessageIter`, so don't have an opinion on that. What would be the return type for `NlSockHanle::recv`? As far as the return type is easily handable, I'm completely fine with your proposal.

**Comment by jbaublitz** [comment]
The return type would remain the same.

**Comment by Kloenk** [comment]
What if the kernel sends an Ack and a Payload? Do you have to to call `recv` twice, or how would you design this?

**Comment by jbaublitz** [comment]
You would have to call `.recv()` twice and to simply iterate until the end, you would use `NlMessageIter`.

**Comment by jbaublitz** [comment]
@Kloenk Once you give this the okay, I'll merge! I've tested it extensively and think this should work now.

File: src/socket.rs [hunk 2]
```diff
@@ -394,21 +394,22 @@ impl NlSocketHandle {
     /// Convenience function for resolving a [`str`] containing the
     /// generic netlink family name to a numeric generic netlink ID.
     pub fn resolve_genl_family(&mut self, family_name: &str) -> Result<u16, NlError> {
+        let mut res = Err(NlError::new(format!(
+            "Generic netlink family {} was not found",
```
Comment (Kloenk) [inline src/socket.rs:398]: I like to do stuff like this in `'{}'`. but I guess that's a personal favourite, I'm fine with either way

**Review by Kloenk — APPROVED** [review]
Did not test it, but so far LGTM

**Comment by jbaublitz** [comment]
@Kloenk Okay sounds good! Let me know when you test it and then I'll merge and put out the new release.

**Comment by jbaublitz** [comment]
Or if you want me to test it, just tell me what packet I should send to the `nl80211` family.

**Comment by Kloenk** [comment]
Will test it right now, but have to build a new kernel (old build is not cached anymore and the nl80211 stuff I'm testing uses EXPERT features of the kernel)

**Comment by Kloenk** [comment]
The tests in my project are succeeding. So works for my current needs. 
LGTM

will this mean a release today? (asking if I should commit the version from git or wait for a tag on crates.io)

**Comment by jbaublitz** [comment]
This does mean a release today. I'll do it right when I get home.

**Pull request merged**